### ISSUE TITLE
Comments no longer work in appsettings.json

### ISSUE THREAD

**Issue by Cynicszm** [fededim/Fededim.Extensions.Configuration.Protected#26]
Hi,

As far as I understand comments are explicitly allowed in appsettings.json (https://github.com/dotnet/AspNetCore.Docs/issues/26254#issuecomment-1169775262). 

Version 1.0.7 had no problems but 1.0.9 now throws the exception below when calling ProtectFiles on a file with comments e.g. 
// some comment

Removing all comments works.  

```
System.Text.Json.JsonReaderException
  HResult=0x80131500
  Message='/' is an invalid start of a property name. Expected a '"'. LineNumber: 65 | BytePositionInLine: 2.
  Source=System.Text.Json
  StackTrace:
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes) in /_/src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.cs:line 315
   at System.Text.Json.Utf8JsonReader.ConsumeNextToken(Byte marker) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs:line 1786
   at System.Text.Json.Utf8JsonReader.ConsumeNextTokenOrRollback(Byte marker) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs:line 1689
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment() in /_/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs:line 833
   at System.Text.Json.Utf8JsonReader.Read() in /_/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs:line 279
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, JsonReaderOptions readerOptions, MetadataDb& database, StackRowStack& stack) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs:line 921
   at System.Text.Json.JsonDocument.ParseUnrented(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, JsonTokenType tokenType) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs:line 752
   at System.Text.Json.JsonDocument.ParseValue(ReadOnlyMemory`1 json, JsonDocumentOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs:line 315
   at System.Text.Json.Nodes.JsonNode.Parse(String json, Nullable`1 nodeOptions, JsonDocumentOptions documentOptions) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonNode.Parse.cs:line 80
   at Fededim.Extensions.Configuration.Protected.JsonFileProtectProcessor.ProtectFile(String rawFileText, Regex protectRegex, Func`2 protectFunction) in Fededim.Extensions.Configuration.Protected\JsonFileProtectProcessor.cs:line 13
   at Fededim.Extensions.Configuration.Protected.ConfigurationBuilderExtensions.ProtectFiles(IDataProtector dataProtector, String path, String searchPattern, SearchOption searchOption, String protectRegexString, String protectedReplaceString, Boolean backupOriginalFile) in Fededim.Extensions.Configuration.Protected\ConfigurationBuilderExtensions.cs:line 58
```

**Comment by fededim** [comment]
Yep, you are right again, forgot to set default parsing and serialization options allowing comments in JSON files after rework. I have fixed it now (and also added comments to json files inside Fededim.Extensions.Configuration.Protected.ConsoleTest in order to always test them), released a version 1.0.10 which should be fine, please update and let me know if everything is fine.

For XML there is no such issue since the XML standard unlike JSON allows comments.

**Comment by Cynicszm** [comment]
Thanks for the quick work. The fix works when loading an appsettings.json file with comments but then strips them out of the output file.

**Comment by fededim** [comment]
**Yep, you are right that because the JSON standard does not support comments (incredibly) as XML does.**

The System.Text.Json is compliant to the standard but allowed in the past the handling of comments inside JSON files through the serialization option **JsonCommentHandling.Allow**. Unluckily since NET Core 3.1 this option has been deprecated and raises always an exception in MS code (see [JsonDocumentOptions source code](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocumentOptions.cs#L27-L38"/>) and [JsonDocument source code](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L1083-L1094)); NET now only supports reading and skipping of comments inside JSON files but not writing. So this is the reason why they are not preserved in the output encrypted files.

**As a workaround to solve this issue however I implemented another FileProtectProcessor called JsonWithCommentsFileProtectProcessor** (which is hacky, it basically replaces string tokens inside json files decoding and re-enconding their values). It is not active by default, if you want to activate it you have to call the static method ConfigurationBuilderExtensions.UseJsonWithCommentsFileProtectOption();

You can find all these changes inside the new version 1.0.11 of the package (should be already available on Nuget.org), I'll update the article in the following days.

Again let me know if there is something which does not work.

**Comment by fededim** [comment]
@Cynicszm please let me know if the fix works. Moreover I have just released the new version 1.0.12 which has:

- Improvement: Allow encryption/decryption to be pluggable with providers using a new interface IProtectProvider. Therefore all DataProtectionAPI dependencies have been moved to a new package Fededim.Extensions.Configuration.Protected.DataProtectionAPI, you can just use this one which requires Fededim.Extensions.Configuration.Protected.
- Bugfix: Fixed a bug with the subPurpose section of the regexs which could lead to a greedy match instead of a lazy one.

To use the Data Protection API as before you can reference the new package Fededim.Extensions.Configuration.Protected.DataProtectionAPI (it includes Fededim.Extensions.Configuration.Protected) or if you wish to implement custom encryption/decryption (to be secure this requires knowledge!) you can use Fededim.Extensions.Configuration.Protected and implement your own logic using the IProtectProvider interface.

**Comment by Cynicszm** [comment]
I've tested with 1.0.12 and the fix appears to be working, No more disappearing comments when using
`ConfigurationBuilderExtensions.UseJsonWithCommentsFileProtectOption()` 

Thanks

**Comment by fededim** [comment]
Ok, I released version 1.0.13 a few hours ago, it's just a renaming of methods and has some minor improvements, I'll leave this bug open till the end of the month, should you find something which does not work please let me know.

**Comment by fededim** [comment]
I have just released version 1.0.14 of Fededim.Extensions.Configuration.Protected and version 1.0.2 of Fededim.Extensions.Configuration.Protected.DataProtectionAPI, it includes extensive unit testing for all configuration sources (10000 entries for each one). Hopefully this version should be the stable one, I close the issue.

![image](https://github.com/fededim/Fededim.Extensions.Configuration.Protected/assets/8364158/4864bec6-36e2-44b1-816e-c30e1992a788)
